File: local-network/deploy/src/waves-txs.ts
import * as wt from '@waves/waves-transactions';
import * as s from './common-settings';

// cc - Chain contract
// 1. Deploy chain contract script
export function mkCcDeploy(script: string) {
  return wt.setScript(
    {
      chainId: s.chainId,
      fee: 2_900_000,
      script: script
    },
    { privateKey: s.chainContract.privateKey }
  )
}

// 2. Setup chain contract
export function ccSetup(elGenesisBlockHashHex: string) {
  return wt.invokeScript(
    {
      chainId: s.chainId,
      fee: 100_500_000,
      dApp: s.chainContract.address,
      call: {
        function: "setup",
        args: [
          {
            // genesisBlockHashHex
            type: "string",
            value: elGenesisBlockHashHex.substring(2)
          },
          {
            // minerRewardInGwei
            type: "integer",
            // 2_000_000_000 Gwei = 2_000_000_000*10^9 Wei = 2*10^18 Wei = 2 UNIT0 for epoch, 
            // see bridge.sol for conversion details
            value: 2_000_000_000
          }
        ]
      }
    },
    { privateKey: s.chainContract.privateKey }
  )
}

// 3 & 4. Join EL miners on CL
export const ccMinerJoin1 = wt.invokeScript(
  {
    chainId: s.chainId,
    fee: 500_000,
    dApp: s.chainContract.address,
    call: {
      function: "join",
      args: [
        {
          // Reward address
          type: "binary",
          // 0x7dbcf9c6c3583b76669100f9be3caf6d722bc9f9
          value: "base64:fbz5xsNYO3ZmkQD5vjyvbXIryfk="
        }
      ]
    }
  },
  { privateKey: s.wavesMiner1.privateKey }
);

export const ccMinerJoin2 = wt.invokeScript(
  {
    chainId: s.chainId,
    fee: 500_000,
    dApp: s.chainContract.address,
    call: {
      function: "join",
      args: [
        {
          // Reward address
          type: "binary",
          // 0xcf0b9e13fdd593f4ca26d36afcaa44dd3fdccbed
          value: "base64:zwueE/3Vk/TKJtNq/KpE3T/cy+0="
        }
      ]
    }
  },
  { privateKey: s.wavesMiner2.privateKey }
);

// Setup is done

// CL -> EL withdraw
export function mkC2ETransfer(senderPrivateKey: string, destElAddressHex: string, amount: bigint, tokenIdB58: string, timestamp?: number) {
  return wt.invokeScript(
    {
      chainId: s.chainId,
      fee: 500_000,
      dApp: s.chainContract.address,
      call: {
        function: "transfer",
        args: [
          { type: 'string', value: destElAddressHex.slice(2) },
        ]
      },
      payment: [
        {
          amount: Number(amount),
          assetId: tokenIdB58
        }
      ],
      timestamp: timestamp || Date.now()
    },
    { privateKey: senderPrivateKey }
  )
}

// EL -> CL transfer
export function mkE2CTransfer(senderPrivateKey: string, blockHash: string, merkleTreeProofs: Buffer[], withdrawIndex: number, amount: bigint) {
  return wt.invokeScript(
    {
      chainId: s.chainId,
      fee: 500_000,
      dApp: s.chainContract.address,
      call: {
        function: "withdraw",
        args: [
          { type: 'string', value: blockHash.slice(2) },
          { type: 'list', value: merkleTreeProofs.map(x => ({ type: 'binary', value: x.toString('base64') })) },
          { type: 'integer', value: withdrawIndex },
          { type: 'integer', value: Number(amount) }
        ]
      }
    },
    { privateKey: senderPrivateKey }
  )
}

File: local-network/deploy/local/waves_txs.py
# TODO move to classes

import sys
from logging import Logger
from time import sleep

from pywaves import pw
from units_network.common_utils import hex_to_base64


def force_success(log: Logger, r, text, wait=True, pw=pw):
    if not r or "error" in r:
        log.error(f"{text}: {r}")
        sys.exit(1)

    if wait:
        id = r["id"]
        wait_for_txn(id, pw)
        log.info(f"{id} confirmed")


def wait_for_txn(id, pw=pw):
    while True:
        tx = pw.tx(id)
        if "id" in tx:
            return tx
        sleep(2)


def clean_hex_prefix(hex: str) -> str:
    return hex[2:] if hex.startswith("0x") else hex


# cc - Chain contract
# 1. Deploy chain contract script
def cc_set_script(chain_contract: pw.Address, script: str):
    return chain_contract.setScript(script, txFee=3_200_000)


# 2. Setup chain contract
def cc_setup(chain_contract: pw.Address, el_genesis_block_hash_hex: str):
    return chain_contract.invokeScript(
        dappAddress=chain_contract.address,
        functionName="setup",
        params=[
            {"type": "string", "value": clean_hex_prefix(el_genesis_block_hash_hex)},
            {
                # minerRewardInGwei
                "type": "integer",
                # 2_000_000_000 Gwei = 2_000_000_000*10^9 Wei = 2*10^18 Wei = 2 UNIT0 for epoch,
                # see bridge.sol for conversion details
                "value": 2_000_000_000,
            },
        ],
        txFee=100_500_000,
    )


# 3. Join EL miners on CL
def cc_join(
    chain_contract_address: str, sender: pw.Address, el_reward_address_hex: str
):
    return sender.invokeScript(
        dappAddress=chain_contract_address,
        functionName="join",
        params=[
            {
                "type": "binary",
                "value": f"base64:{hex_to_base64(clean_hex_prefix(el_reward_address_hex))}",
            },
        ],
        txFee=500_000,
    )
